rskj-core/src/test/java/co/rsk/peg/performance/ReceiveHeaderTest.java

```diff
@@ -0,0 +1,134 @@
+    }
+
+    @Test
+    public void ReceiveHeaderTest() throws VMException {
+        ExecutionStats stats = new ExecutionStats("receiveHeader");
+        receiveHeader_success(1000, stats);
+        receiveHeader_block_already_saved(500, stats);
+        Assert.assertTrue(BridgePerformanceTest.addStats(stats));
+    }
+
+    private void receiveHeader_success(int times, ExecutionStats stats) throws VMException {
+        int minCentsBtc = 10;
+        int maxCentsBtc = 100;
+
+        BridgeStorageProviderInitializer storageInitializer = generateInitializer(minCentsBtc, maxCentsBtc, true);
+
+        executeAndAverage(
+                    "receiveHeader_success",
+                    times,
+                    getABIEncoder(true),
+                    storageInitializer,
+                    Helper.getZeroValueRandomSenderTxBuilder(),
+                    Helper.getRandomHeightProvider(10),
+                    stats,
+                    (environment, executionResult) -> {
+
+                        // Working fine.
+                        int result = new BigInteger(executionResult).intValue();
+                        Assert.assertEquals(0, result);
+
+                        // Best Block is the new block added.
+                        btcBlockStore = new RepositoryBtcBlockStoreWithCache(BridgeRegTestConstants.getInstance().getBtcParams(), (Repository) environment.getBenchmarkedRepository(), new HashMap<>(), PrecompiledContracts.BRIDGE_ADDR);
+                        Sha256Hash bestBlockHash = null;
+                        try {
+                            bestBlockHash = btcBlockStore.getChainHead().getHeader().getHash();
+                        } catch (BlockStoreException e) {
+                            Assert.fail(e.getMessage());
+                        }
+                        Assert.assertEquals(expectedBlock.getHash(), bestBlockHash);
+                    }
+            );
+        }
+
+    private void receiveHeader_block_already_saved(int times, ExecutionStats stats) throws VMException {
+        int minCentsBtc = 10;
+        int maxCentsBtc = 100;
+
+        BridgeStorageProviderInitializer storageInitializer = generateInitializer(minCentsBtc, maxCentsBtc, false);
+
+        executeAndAverage(
+                "receiveHeader_success",
+                times,
+                getABIEncoder(false),
+                storageInitializer,
+                Helper.getZeroValueRandomSenderTxBuilder(),
+                Helper.getRandomHeightProvider(10),
+                stats,
+                (environment, executionResult) -> {
+                    int result = new BigInteger(executionResult).intValue();
+                    Assert.assertEquals(-4, result);
+                }
+        );
+    }
+
+    private ABIEncoder getABIEncoder(boolean createBlock) {
+            return (int executionIndex) -> {
+                if (createBlock) {
+                    expectedBlock = Helper.generateBtcBlock(lastBlock);
```

Comment:
instead of relying on a class level field I recommend refactoring this to do the following:
- expect a Block parameter in `getABIEncoder`
- for the `receiveHeader_success` use this logic `Helper.generateBtcBlock(lastBlock)` to get the next block to inform.
- for the `receiveHeader_block_already_saved` get the bestBlock in the repository and use this as the parameter for `getABIEncoder`

Reply:
We don't need to execute cases where the block save height is not the last one?